addons/product_margin/tests/test_product_margin.py

```diff
@@ -0,0 +1,118 @@
+            'type': 'in_invoice',
+            'company_id': self.ref('base.main_company'),
+            'name': 'invoice to supplier',
+            'invoice_line_ids': [(0, 0, {
+                        'name': self.Product.name,
+                        'product_id': self.Product.id,
+                        'quantity': 4.0,
+                        'account_id': account_expense,
+                        'uom_id': self.ref('product.product_uom_unit'),
+                        'price_unit': 450.00
+                    })],
+        }).action_invoice_open()
+
+        # Define Customer Invoice
+        self.AccountInvoice.create({
+            'partner_id': self.ref("base.res_partner_1"),
+            'name': 'invoice to client',
+            'type': 'out_invoice',
+            'company_id': self.ref('base.main_company'),
+            'account_id': account_receivable,
+            'invoice_line_ids': [(0, 0, {
+                        'name': self.Product.name,
+                        'product_id': self.Product.id,
+                        'quantity': 20.0,
+                        'account_id': account_revenue,
+                        'uom_id': self.ref('product.product_uom_unit'),
+                        'price_unit': self.Product.list_price
+                    })],
+        }).action_invoice_open()
+
+        self.AccountInvoice.create({
+            'partner_id': self.ref("base.res_partner_2"),
+            'name': 'invoice to client',
+            'type': 'out_invoice',
+            'company_id': self.ref('base.main_company'),
+            'account_id': account_receivable,
+            'invoice_line_ids': [(0, 0, {
+                        'name': self.Product.name,
+                        'product_id': self.Product.id,
+                        'quantity': 10.0,
+                        'account_id': account_revenue,
+                        'uom_id': self.ref('product.product_uom_unit'),
+                        'price_unit': 550.00
+                    })],
+        }).action_invoice_open()
+
+        result = self.Product._compute_product_margin_fields_values(self)
+
+        # Sale turnover ( Quantity * Unit price)
+        sale_turnover = ((20.0 * 750.00) + (10.0 * 550.00))
+
+        # Expected sale (Total quantity * Sale price)
+        sale_expected = (750.00 * 30.0)
+
+        # Purchase total cost (Quantity * Unit price)
+        purchse_total_cost = ((10.0 * 300.00) + (4.0 * 450.00))
+
+        # Purchase normal cost ( Total quantity * Cost price)
+        purchse_normal_cost = (14.0 * 500.00)
+
+        total_margin = sale_turnover - purchse_total_cost
+        expected_margin = sale_expected - purchse_normal_cost
+
+        # Check total margin
+        self.assertEqual(result[self.Product.id]['total_margin'], total_margin, "Wrong Total Margin.")
+
+        # Check expected margin
+        self.assertEqual(result[self.Product.id]['expected_margin'], expected_margin, "Wrong Expected Margin.")
```

Comment:
These assertion messages don't really bring anything to the table, it's rather clear from the good variable names that we're making assertions based on total_margin and expected_margin.